### ISSUE TITLE
expression with remainder function

### ISSUE THREAD

**Issue by tmIdiada** [esmini/esmini#319]
Hi,

I'm getting the following:
```
Loaded OpenDRIVE: ../xodr/straight500.xodr
Expr ${(($offset_percentage + 100) % 100) / 100 * $targetWidth_m} = ((75 + 100) % 100) / 100 * 1.8 = 1.3500000000
```
So, remainder of 175 and 100 seems here to be 75, the same like in python 175 % 100. But, if I call in python from math module the remainder function, then I get -25. And exactly this remainder behavior I would expect here, but maybe I'm wrong.

How you would design the above expression to get the following behavior:
```
offset percentage input and offset regarding center of lane as output
100% -> 0
50% -> -50 or -0.5
75% -> -25 or -0.25
-50% -> 50 or 0.5
-75% -> 25 or 0.25
```

Can you help me to understand which behavior in OSC for remainder here is defined and if esmini is doing the right thing?

P.S.: I don't want to include operations which are supported by esmini, but not part of user guide, or if there is any risk that the operation is not supported by other tools which are OSC v1.1 compliant. For example I've seen you are supporting larger than and smaller than operations, that way I could design max or min or sign functions here which would realize the expected behavior. But I don't see ">" or "<" operators in user guide here.

**Comment by tmIdiada** [comment]
Solved by using the following formula instead
```
${($offset_percentage - round($offset_percentage / 100) * 100) / 100 * $targetWidth_m}
```
Works as expected and even 25% is not foreseen with round instead of ceil it works exactly like -75% and -25% works like 75%.

**Comment by eknabevcc** [comment]
Tricky question. Glad that you found a solution.
The c++ operation used is fmod: https://cplusplus.com/reference/cmath/fmod/

**Comment by tmIdiada** [comment]
> Tricky question. Glad that you found a solution. The c++ operation used is fmod: https://cplusplus.com/reference/cmath/fmod/

Please check maybe together with ASAM if the remainder here is understood by all parties as unsigned remainder, then you are correct using fmod. At least Javascript and python math.remainder are defining it as signed remainder. For me not clear defined.

**Comment by eknabevcc** [comment]
Thanks, didn't see the nice EUNCAP image until now - good example and use case 👍

Yes, it seems that the operator needs to be clarified. 

I see multiple common implementations of modulo/remainder (some picked from https://en.wikipedia.org/wiki/Modulo_operation):
C++ % and fmod: truncate division (always towards zero, just removing decimals)
C++ remainder: round division (closest integer)
Python %: floor division (always down)
Python math.fmod: truncate division (always towards zero, just removing decimals)
Python math.remainder: round division (closest integer)

I found this statement in [OpenSCENARIO User guide](https://www.asam.net/index.php?eID=dumpFile&t=f&f=4908&token=ae9d9b44ab9257e817072a653b5d5e98ee0babf8#_expressions):

> For the definition of the arithmetic operators [[13](https://www.asam.net/index.php?eID=dumpFile&t=f&f=4908&token=ae9d9b44ab9257e817072a653b5d5e98ee0babf8#IEEE_754-2019)] shall be used. It is noteworthy that % is the remainder operator, not the modulo operator. For integers greater zero, both operators give equal results. Furthermore, the arguments of the operators round, floor, ceil, and pow are given behind the operator name and need to be surrounded by parentheses.

[13] refer to [IEEE 754](https://en.wikipedia.org/wiki/IEEE_754)

So, along with your understanding, the true remainder operator seems to be the one intended. However, the statement "For integers greater zero, both operators give equal results" reveals that it might be a mistake involved. I mean, maybe the "Python style" modulo operator was actually the intended one. 

Anyway, the conflicting info should be corrected or at least clarified in next OpenSCENARIO version (the IEEE remainder gives e.g. -25 for 175 % 200, i.e. NOT 175 which is the case for both variants of the modulo operator)

To conclude, I can't see what is the correct or even "best" choice for esmini. I think at least C++ style fmod is commonly understood. And a change in itself is not to prefer. So maybe we stick with current implementation until sorted out. I don't have a strong opinion myself, but I suspect that negative results might confuse some users at least for two positive numbers (and since the '%' is typically referred to as modulo operator).

What would be your opinion for 1. esmini implementation and 2. OpenSCENARIO v1.3?

**Comment by tmIdiada** [comment]
Hi,

the picture I have added afterwards yesterday evening as well the corrected offset calculation, First I had only the width of the GVT which holds only for the 100% and the 50% cases, but not for other percentage values in between, now it should work for all, but anyone else interested is very welcome to recheck it and correct it.

Which solution is preferred from my side, I would more need sign, max, min, >, <, abs and maybe sin and cos functions and operators than any other version for modulo or remainder.

From general user point of view the remainder like fmod in C++ is better, but from expert point of view one has definitely more flexibility using the signed remainder function than the current implementation, you can see it on the function expression in my first comment. But I had for sure never the idea to realise it that way if there would be not this sentence in user guide that not the modulo operator is meant and the remainder is taken and a short google search afterwards. Maybe the general understanding of % operator is better here, because the behavior of the signed remainder is sometimes hard to predict and one has to do a lot of tests as programmer.

Kind regards,

TM

**Comment by eknabevcc** [comment]
Good input!

**Comment by tmIdiada** [comment]
BTW: Keep attention for round function expression

> Round to nearest, ties to even is the default for binary floating point and the recommended default for decimal. Round to nearest, ties to away is only required for decimal implementations
> 
> Example of rounding to integers using the IEEE 754
> Mode | +11.5 | +12.5 | −11.5 | −12.5
> to nearest, ties to even | +12.0 | +12.0 | −12.0 | −12.0
> to nearest, ties away from zero | +12.0 | +13.0 | −12.0 | −13.0

My previous implementation with round function instead of ceil worked for the 50% case, but the default should be to nearest, ties to even, maybe you are also wrong here, otherwise I'm wondering why my 50% case worked.

**Comment by eknabevcc** [comment]
**Regarding `round` function:**
esmini use the standard C/C++ implementation which seems to always tie away from 0. From https://cplusplus.com/reference/cmath/round: "Returns the integral value that is nearest to x, with halfway cases rounded away from zero."
Personally I think that's less confusing and more useful than tie to even. So, even if not compliant with IEEE 754, I'll keep current implementation for now.

**Regarding additional functions**
Good suggestion. Since implementation was easy, I added support for some additional functions. See [User guide - Expressions](https://esmini.github.io/#_expressions).

I also created a suggestion for OpenSCENARIO 1.3. But of course, no guarantee that it will be accepted. So for now, you should be aware that these functions will probably not work in other simulators.

Ok to close issue now you think?

**Comment by tmIdiada** [comment]
Thank you a lot for the new functions, you have a typo in modulo/rem**A**inder.

**Comment by tmIdiada** [comment]
FYI (from official ASAM OSC point of view your implementation seems to be wrong for remainder, for the round function I would say as well)
**The following answer I got from ASAM CTO (translated by google translate):**

> In Python, % is actually the modulo operator (in C, % is the remainder operator, by the way). Python 3.7 added the IEEE 754 remainder operator (**this is the correct one for OSC**). Since we didn't do any cross-tests, I can't guarantee that the implementation is correct for all tool vendors, but I'm relatively sure that it is the same for the common ones.
> https://docs.python.org/3.7/whatsnew/3.7.html
> math.remainder
> Return the IEEE 754-style remainder of x with respect to y. For finite x and finite nonzero y, this is the difference x - n*y, where n is the closest integer to the exact value of the quotient x / y. If x / y is exactly halfway between two consecutive integers, the nearest even integer is used for n. The remainder r = remainder(x, y) thus always satisfies abs(r) <= 0.5 * abs(y).
> modulo %
> This is m = x - n*y where n is the floor(x/y), so 0 <= m < y instead of abs(r) <= 0.5 * abs(y) for the remainder.
> modulo(2.7, 1) = 0.7
> remainder(2.7, 1) = -0.3
> 
>

**Comment by eknabevcc** [comment]
Thanks for info. 

I think there's still confusion. The "%" operator in C/C++ is only defined for integers, why I don't think it correlates to the IEEE 754 remainder operation. 

I don't have a strong opinion here. Do you see better use for the `remainder` operation? I could then consider changing to: https://en.cppreference.com/w/cpp/numeric/math/remainder - it seems to be aligned with IEEE 754. 

I see at least the nice use case to map a number to an interval [-N/2, N/2], e.g. angle to [-pi, pi].

Then regarding `round`. I see there are [multiple options for rounding rules](https://en.wikipedia.org/wiki/IEEE_754#Rounding_rules) in IEEE 754. Of which the C++ (I assume) [Round to nearest, ties away from zero](https://en.wikipedia.org/wiki/Rounding#Round_half_away_from_zero) is one option. But to my understanding the [Round to nearest, ties to even](https://en.wikipedia.org/wiki/Rounding#Round_half_to_even) is default. In C++ it seems to be implemented as https://en.cppreference.com/w/cpp/numeric/math/rint. 

So, yes, it would be easy to change mapping of these two operators in esmini. 

What is your opinion of today? Do you see reason to change, in addition to complying to IEEE? 

Personally I think the tie-to-even could cause some surprises :) But the remainder is potentially useful. 

One compromise can be to change the `remainder` implementation but wait with changing the `round` until discussed in next OpenSCENARIO 1.X project. What do you think?

**Comment by tmIdiada** [comment]
In my opinion esmini should strictly fulfill all OSC definitions, this would mean here 100% coverage of IEEE754 with remainder and round function, although both implementations/definitions are really not my favorite or better I don't like these choices at all.

Here I like more the current implementations of esmini, but we know the definitions from ASAM and IEEE and esmini should operate here like all other tools following the standard.

I'm not happy with it, but in my opinion esmini should surprise here the users and strictly refer to OSC user guide and the linked IEEE standard.

And in following OSC meetings hopefully at least one member will suggest to use the more obvious solutions at least for the round function ties away from zero instead of ties to even, but if a standard uses this definition as default I see no chance why ASAM should not follow this standard.

**Comment by eknabevcc** [comment]
Now there's also a branch along with the "compromise" approach: [update_operators](https://github.com/esmini/esmini/tree/update_operators)
commit [97a75cb](https://github.com/esmini/esmini/commit/97a75cba69866a862a06808f3b2502166eb9a84e) do the change

You're welcome to try it out if possible.

**Comment by eknabevcc** [comment]
We wrote simultaneously, so I didn't see yours before my message regarding branch. 

Ok, I also changed the round implementation (commit [9e6856e](https://github.com/esmini/esmini/commit/9e6856e838e906f001cd37bfcf43848f53da2c38)). So both `%` and `round` should be according to IEEE 754 default in the branch.

I'll discuss internally and see if I find other opinions before merging to master.

It would also be great if you could test and confirm expected behavior.

**Comment by tmIdiada** [comment]
Hi,

round and remainder work as expected! Tested with acc-test and defined remainder and round for offset of target vehicle.

If you decide finally to take the official definitions, then please include these functions to some of your xosc and your test procedure with the corresponding baseline (new formulas should result in new baseline with one additional line, even the value would not change). For example if you use somewhere offset=2, then please replace it by offset="${round(2.5)}" and at some other position offset="${round(1.5)}" or offset="${7 % 5}" or offset="${-(8 % 5)}". For sure, you can also replace these numbers by parameters set in parameter declarations.

**Comment by eknabevcc** [comment]
Unit test cases updated, and also a few new ones added:
[Unittest/ScenarioEngine_test.cpp starting at line 310](https://github.com/esmini/esmini/blob/206491631e3ef363e8f11c9f5ccbd60e826de270/EnvironmentSimulator/Unittest/ScenarioEngine_test.cpp#L310)

And User guide has been updated (but not published on esmini.github.io yet):

![image](https://user-images.githubusercontent.com/29146839/196387842-ec7351e9-0d4b-4d6c-afb1-5e84791d75c4.png)

**Comment by eknabevcc** [comment]
Let's go for it. I included these updates in release [v2.26.8](https://github.com/esmini/esmini/releases/tag/v2.26.8). 
We'll see if we get some feedback. But I agree with you that the aim should be to align with the standard.
